### PR TITLE
Adding pre-commit tool config plus cmake-format

### LINKED ISSUE #3591: Add and enforce CMake formatting rules
We should have a .cmake-format file, and enforce it in the CI. We might also consider checking
if there are source files that are not mentioned in the build system and warn about that (not sure about this one).

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Fixes #3591

Depends on https://github.com/espressomd/docker/pull/158

Description of changes:
- pulled out tool version checking to separate files
- added configuration file for [pre-commit](https://pre-commit.com)
- added cmake-format configuration and style fix

### CODE REVIEW

File: maintainer/format/autopep8.sh
Comment (jngrad) [inline maintainer/format/autopep8.sh:19]: here and in the other new .sh files: should we copy-paste the license header from `maintainer/CI/fix_style.sh`?

File: requirements.txt
Comment (jngrad) [inline requirements.txt:31]: should we pin `cmake-format==0.6.9` just like we pinned `autopep8` and `pycodestyle`?

File: maintainer/format/autopep8.sh
Comment (KaiSzuttor) [reply maintainer/format/autopep8.sh:19]: yes probably

File: maintainer/format/autopep8.sh
Comment (jngrad) [inline maintainer/format/autopep8.sh]: Do we really need Bash? The original file was converted pure Dash ([`984b061`](https://github.com/espressomd/espresso/commit/984b061645a80f5e184eabf077a43ffddb0c26b8#diff-78719c369a7782258dc4e7743964d748)) to improve interoperability and prevent the re-introduction of bashisms.

File: doc/tutorials/01-lennard_jones/CMakeLists.txt
Comment (jngrad) [inline doc/tutorials/01-lennard_jones/CMakeLists.txt]: The majority of the changes looks very good, but in tutorials, the changes look inconsistent: the `configure_tutorial_target` command is nicely condensed while the `nb_export` command is split over too many lines; can't we have it condensed over two lines only? This would be below the 80 chars limit:
```cmake
nb_export(TARGET tutorial_04 SUFFIX "1"
          FILE "04-lattice_boltzmann_part1.ipynb" HTML_RUN)
```

File: src/core/unit_tests/CMakeLists.txt
Comment (jngrad) [inline src/core/unit_tests/CMakeLists.txt]: This formatting should not happen; please revert this change and add the missing `#` symbol in empty lines of the license header.

File: src/core/unit_tests/CMakeLists.txt
Comment (jngrad) [inline src/core/unit_tests/CMakeLists.txt]: One word by line is overzealous.

File: src/shapes/unit_tests/CMakeLists.txt
Comment (jngrad) [inline src/shapes/unit_tests/CMakeLists.txt]: overzealous

File: src/utils/tests/CMakeLists.txt
Comment (jngrad) [inline src/utils/tests/CMakeLists.txt]: overzealous

File: testsuite/python/CMakeLists.txt
Comment (jngrad) [inline testsuite/python/CMakeLists.txt]: overzealous

**Review by jngrad — CHANGES_REQUESTED** [review]

File: doc/tutorials/01-lennard_jones/CMakeLists.txt
Comment (jngrad) [reply doc/tutorials/01-lennard_jones/CMakeLists.txt]: all the other tutorials are still affected

Comment (KaiSzuttor) [reply doc/tutorials/01-lennard_jones/CMakeLists.txt]: I think you looked at outdated files.

File: doc/tutorials/04-lattice_boltzmann/CMakeLists.txt
Comment (jngrad) [inline doc/tutorials/04-lattice_boltzmann/CMakeLists.txt:31]: overzealous

File: doc/tutorials/05-raspberry_electrophoresis/CMakeLists.txt
Comment (jngrad) [inline doc/tutorials/05-raspberry_electrophoresis/CMakeLists.txt:9]: overzealous

**Review by jngrad — CHANGES_REQUESTED** [review]

File: doc/tutorials/04-lattice_boltzmann/CMakeLists.txt
Comment (KaiSzuttor) [reply doc/tutorials/04-lattice_boltzmann/CMakeLists.txt:31]: that is a matter of taste. If you have a long list it might be more readable to have a column.
Im not sure if there is a global sweet-spot and we have to live with something like that here.
If I further increase the value for the maximum elements you will get more of unreadable 
horizontal lists.

**Comment by fweik** [comment]
Just my two cents: I'm not sure that adding `pre-commit` is worth it over just calling the shell scripts. It's just one more thing to know. @jngrad if you decide to keep it, make sure that @KaiSzuttor properly documents it in the CI documentation.

**Comment by jngrad** [comment]
@fweik once this is merged, we won't be able to call the shell scripts anymore, because the `find` commands that detect the file extensions were replaced by `pre-commit`. Everyone will have to install `pre-commit` to run the style scripts. I don't really mind, because I don't use `fix_style.sh` anyway; I wrote my own implementation of `pre-commit` almost two years ago (`git-style`) and will keep using it because it has more features.

**Comment by fweik** [comment]
@jngrad yes, the alternative is to keep the find expressions.

**Comment by KaiSzuttor** [comment]
Btw. you already have to pip install the autopep8 package, the pycodestyle package and the cmake-format package due to the pinned versions. You will most probably do that via ```pip install -r requirements.txt```. If so, you will automagically not notice the change at all.

**Review by jngrad — APPROVED** [review]
ok from my side

**Comment by KaiSzuttor** [comment]
interesting
``` bash
Traceback (most recent call last):
24   File "/home/espresso/.local/bin/pre-commit", line 7, in <module>
25     from pre_commit.main import main
26   File "/home/espresso/.local/lib/python3.5/site-packages/pre_commit/main.py", line 189
27     version=f'%(prog)s {C.VERSION}',
28                                   ^
29 SyntaxError: invalid syntax
30 Passed style check
```

**Comment by jngrad** [comment]
@KaiSzuttor We shouldn't delay #3421 any longer and start moving towards Python 3.6 every time our workflow breaks. espressomd/docker#160 moves the style check packages to the Ubuntu 20.04 image, it'll only be a matter of updating the yml file here to use Ubuntu 20.04 image instead of Ubuntu 16.04.

**Comment by KaiSzuttor** [comment]
but, do you understand where the error comes from?

File: maintainer/CI/fix_style.sh
Comment (jngrad) [inline maintainer/CI/fix_style.sh:33]: I think dash and bash return the error code of the last command, so if this line fails due to a Python3.6-specific f-string (Ubuntu 16.04 uses Python3.5), the shell script will keep going to the next instructions and return 0, unless we add the following:
```suggestion
pre-commit run --all-files || exit 1
```

Comment (jngrad) [reply maintainer/CI/fix_style.sh:33]: Actually no, don't do that! `pre-commit` returns 1 when it touched a file.

Comment (jngrad) [reply maintainer/CI/fix_style.sh:33]: Is there an older version of `pre-commit` compatible with Python 3.5 that we can use?

**Comment by KaiSzuttor** [comment]
I just realised that the issue is also that the packaged pip silently installs packages that may not be compatible with the python version. If you upgrade pip and try to install pre-commit in the container, you actually get:
``` bash
ERROR: Package 'pre-commit' requires a different Python: 3.5.2 not in '>=3.6.1'
```
Installing an older version would not help because a dependency of pre-commit needs 3.6.

**Comment by mkuron** [comment]
> > Is there an older version of pre-commit compatible with Python 3.5 that we can use?

> Installing an older version would not help because a dependency of pre-commit needs 3.6.

You could explicitly install specific (python 3.5-compatible) versions of the dependencies to circumvent that problem. Of course, that means adding all dependencies to the Dockerfile. pip is a terrible package manager.

**Comment by RudolfWeeber** [comment]
I wonder whether this PR takes too much work from too many people.

 

Personally, I just run JN’s script before committing and that’s that.

 

Do we really want to follow through with this?

 

Or is my impression wrong, that this hasn’t converged to the point of just needing a review?

**Comment by KaiSzuttor** [comment]
> Or is my impression wrong, that this hasn’t converged to the point of just needing a review?

my impression is that sometimes we comment on PRs without having looked at the changes

**Comment by jngrad** [comment]
> I wonder whether this PR takes too much work from too many people.

There wouldn't be so much friction if ICP machines had Clang-9 installed.

> Do we really want to follow through with this?

The `pre-commit` tool removes some complexity from the shell scripts by taking care of the path detection. It's unclear to me how many people will use the commit hook feature. I most likely won't, because I rely a lot on rebase operations and don't want them to be interrupted, nor do I want to work on two clones, one with the hook and one without.

It's also unclear to me why `pre-commit` works so fast. On my previous ICP machine that had twice as less RAM the `fix_style.sh` script would use several GB of RAM and freeze the OS for several minutes (Rudolf and Christoph also experience performance issues). I'd prefer to have `pre-commit` in ESPResSo rather than my tool in ESPResSo, because my tool is untested and quite complex.

**Comment by KaiSzuttor** [comment]
the issues came solely from the default pip version, I fixed it in https://github.com/espressomd/docker/pull/162

**Comment by KaiSzuttor** [comment]
closed for #3622

**Pull request closed without merging**
